### PR TITLE
Complete rework of the ghost and race recorder

### PR DESCRIPTION
**Pull request by Redix**
This PR fixes a lot of issues with the ghost and race recorder, introduces a new ghost file format and implements some pretty useful features.

**A quick overview:**
 - When you cross the start line multiple times, both (ghost and recorder) will only restart if it is a non-solo server
 - If available, both will use the race timer to recognize the start instead of searching for the start line
 - Fastcap support for both
 - The recorder immediately starts when the Tee spawns, so the preparation steps (which are quite important for speedruns) will be included in the demo. If you do not cross the start-line within 20 seconds, it will stop the demo until you really start.
 - Better ghost menu (colors, reloading, deleting and saving ghosts)
 - The ghosts are more resistant against lags (old ones were sometimes completely useless due to small lags)
 - New ghosts files are significantly smaller
 - Cleanup, bugfixes..

**About the new ghost format (version 4/5):**
This format is used by Teerace for over a year now. The code for handling the binary files was moved to the engine. It includes an auto updater which creates a backup of all ghosts and converts them to the new format afterwards. The major differences from the format used by DDNet right now (version 2), are the ability to store multiple types of data, the usage of delta-encoding and a more portable header.

For whatever reason, the ghost stores values for every predicted tick, but without tick information, so lost snapshots can make them unusable. The new code uses the original values from the snapshots including ticks but it can also handle the old ones without. Since hardly any server uses the high bandwidth option this practically reduces the file size.

Like the demo recorder the ghost recorder directly stores the data to a file (every 50 snapshots) instead of writing the whole file at the end of the race. Indeed this can be changed with only a few lines if the old behavior is preferred.

The updater can handle version 2 (DDNet) and 3 (old teerace format, only slightly different from version 2) files. The updating already happens when the files are scanned for generating the list in the menu and not only when you activate them. The change from version 4 to 5 was only needed due to a bug in the implementation, the ghost loader can read both.

Some numbers about the file size: (map: hotrun, both about 30 seconds)

 - Old ghost: 30.4 kB (converted: 10.7 kB)
 - New ghost: 5.4 kB

**One thing about the race recorder:**
The old implementation compared the new file only with the first file it found for the particular map. The new one compares with all related demos and deletes them possibly, so that only the best demo is left. Since DDNet can also store the demos without name, this might also delete demos from other players, that you might have in your directory.
To prevent this I at least check whether the demo contains the player name if `cl_demo_name` is on. 
In my opinion the better solution would be to remove `cl_demo_name` and always use the player name.

### CODE REVIEW

**Comment by def-** [comment]
So we get a new ghost format here and a new demo format from heinrich?

To me this change sounds great. Anyone up for a review?

**Comment by heinrich5991** [comment]
I hope to find some time for this.

**Comment by Learath2** [comment]
The better solution would not be to delete anything that we didn't create ourselves we already have that information in the demos, so we should check that.

File: src/engine/shared/ghost.cpp [hunk 1]
```diff
@@ -0,0 +1,521 @@
+
```
Comment (Learath2) [inline src/engine/shared/ghost.cpp]: ?

Comment (Learath2) [reply src/engine/shared/ghost.cpp]: Should also clean up the file from trailing whitespaces

File: src/engine/client/client.h [hunk 2]
```diff
@@ -381,12 +383,18 @@ class CClient : public IClient, public CDemoPlayer::IListener
...
-	virtual const char* GetCurrentMap();
-	virtual int GetCurrentMapCrc();
-	virtual const char* GetCurrentMapPath();
-	virtual const char* RaceRecordStart(const char *pFilename);
-	virtual void RaceRecordStop();
-	virtual bool RaceRecordIsRecording();
+	const char *GetCurrentMap();
+	const char *GetCurrentMapPath();
+
+	void RaceRecord_GetName(char *pBuf, int Size, int Time = -1);
```
Comment (Learath2) [inline src/engine/client/client.h]: Don't think we do underscores in function names but should check with @heinrich5991 on that :)

File: src/engine/shared/ghost.cpp (hunk 1 above)
Comment (Redix) [reply src/engine/shared/ghost.cpp]: oops...

File: src/engine/client/client.h (hunk 2 above)
Comment (Redix) [reply src/engine/client/client.h]: seems pretty common:
https://github.com/Redix/ddnet/blob/master/src/engine/client/client.h#L356
and the following lines

**Comment by Redix** [comment]
1. In general I agree that storing the data in the filename is not a good solution. But storing it in the demo would also introduce some problems. Each file has to be opened (needs some time when you have a lot of demos) and in my opinion the demo header should not be changed for that purpose. The time is a ddnet/race specific thing and demos are for recording gameplay in general. Scanning the whole file for specific NETMSGs would be too slow. The easiest solution would be to move the race demos to a new directory, so no other demos will be deleted.

2. This should already work. The code might look a bit messy since it has to deal with local and server-side starting behavior (through race timer), but the ghost rendering should only start when the race has not been restarted since the last tick.

3. I don't see the problem with (up to) 20 additional seconds but it can be limited to non-solo servers easily.

4. As long as you do not change the timer during your race, the code should be able to handle it. But in general I would prefer to always send the race time (in the warmup value) and add a client-side option for the display type (not visible, replace gametimer and maybe a separate one located between the broadcast and gametimer). Non ddnet clients can still use the chat command and choose between none and broadcast timer. This should be enough to detect the start but adding a finish NETMSG would be good to not rely on the chat parsing anymore (although it should be kept for backward compatibility).

**Comment by Learath2** [comment]
2. Doesn't work if I compile this PR

**Comment by Redix** [comment]
2. I can't reproduce this. When restarting is on the ghost starts immediately. When it is off for me the ghost starts when the center of the tee does not touch the start line anymore.
4. Disabling the timer during a race should not be a problem right now. For activating I could add a check if gametick and race start-tick match (or are at least close to each other).

**Comment by Redix** [comment]
Short update:
`/timer` should not affect the ghost and recorder anymore.
Race demos are stored in `demos/auto/race` now.

**Comment by Learath2** [comment]
Interesting that you cannot reproduce 2. I just build it and join a map where I have a ghost on, sit on the start line and watch my ghost go race.

**Comment by Redix** [comment]
Just to make sure that I understand you correctly...
You are on a map with restarting (non-solo)
Your tee is inside the start line
The ghost starts but is reset every tick

Timer was on or off?
Official DDNet Server?

**Comment by Learath2** [comment]
Non-solo map, tee is inside the start line, as soon as I leave the start line the ghost weirdly teleports back and goes where it's supposed to be. I think until you leave the start line the ghost should be paused at the 0 mark.

Timer on, official server.

**Comment by Redix** [comment]
I still cannot reproduce it...
Are all maps affected or only a few?
Does it always happen, only sometimes or only one time after you entered the server?

The problem is, that until we receive the snapshot with the race-timer flag set, the client does not know whether the timer is active or not. So until this snapshot arrives, the client has to check for the start itself, based on predicted values. The transition between the local and the server-side starting behavior might cause some problems. 
I changed some things in the corresponding code. Maybe one of this changes fixed the issue. Would be good if you could try again with the latest code.

When the problem only occurs one time after you joined a map, the client cannot identify the server as non-solo. But I could not imagine why this would fail. It is directly linked to the team visualization in the scoreboard. Someone would probably have noticed if this feature was buggy.

**Comment by Learath2** [comment]
All maps affected, It behaves a little different now but still wrong. 

Now when you go in the start line the first time it doesn't start as expected, but leave the start line and go back on it and the ghost won't disappear. Causing it to re-sync onto you when you leave the line again.

**Comment by Redix** [comment]
Actually this was needed to have a smooth transition from local to server-side ghost control. But I found a solution, so that the ghost is now disabled while you touch the start line.

**Comment by def-** [comment]
So, ready to merge? Sounds like something that would be a good addition for DDNet 11.

**Comment by Learath2** [comment]
Behaviour is fine now \o/ I'll give the code one last look and merge in an hour or so.

File: src/engine/client/client.cpp [hunk 3]
```diff
@@ -3583,45 +3586,56 @@ int main(int argc, const char **argv) // ignore_convention
...
 		m_DemoRecorder[RECORDER_RACE].Stop();
 }
 
-bool CClient::RaceRecordIsRecording()
+bool CClient::RaceRecord_IsRecording()
 {
 	return m_DemoRecorder[RECORDER_RACE].IsRecording();
 }
 
+void CClient::GhostRecorder_Start(const char *pFilename, const char *pPlayerName)
```
Comment (Learath2) [inline src/engine/client/client.cpp]: If `IGhostLoader` and `IGhostRecorder` are kernel interfaces, what is the point of having these functions?

**Comment by Redix** [comment]
The things we talked about are fixed now.
I also tried to make `OnNewSnapshot` at least a bit more readable.
A reliable race timer would simplify the function a bit.

**Comment by Learath2** [comment]
Thanks :)
bors r+

**Pull request merged**